classes/mail/Mailer.inc.php

```diff
@@ -0,0 +1,196 @@
+                $variable = '';
+                continue;
+            }
+
+            if (!empty($stack)) {
+                $variable .= $char;
+                continue;
+            }
+
+            $resultString .= $char;
+        }
+
+        if (!empty($stack))
+            throw new Exception('Template variable tag should be closed');
+
+        return $resultString;
+    }
+
+    /**
+     * @param array $data
+     * @param string|null $locale
+     * @return array localized data
+     */
+    protected function localizeData(array $data, string $locale = null) : array
+    {
+        if (is_null($locale))
+            $locale = PKPLocale::getLocale();
+
+        if (!PKPLocale::isLocaleValid($locale))
+            throw new InvalidArgumentException($locale . ' isn\'t recognized as a valid locale');
+
+        return array_map(function ($varValue) use ($locale) {
+
+            // Array values contain localized template data
+            if (is_array($varValue)) {
+                if (array_key_exists($locale, $varValue)) {
+                    return $varValue[$locale];
+                } else {
+                    throw new InvalidArgumentException('Template variables doesn\'t have localization for the ' . $locale . ' locale' );
+                }
+            }
+            return $varValue;
+        }, $data);
+    }
+
+    /**
+     * @param int $key number of char in a string starting from 0
+     * @param string $char
+     * @param array $template array of chars of a template
+     * @return bool
+     * @brief defines an opening tag of a variable
+     */
+    protected function openTags(int $key, string $char, array $template) : bool
+    {
+        if ($char !== self::OPEN_TAG)
+            return false;
+        if ($template[$key+1] === self::DOLLAR_SIGN_TAG)
+            return true;
+        return false;
+    }
+
+    /**
+     * @param string $compiledString the port string that has been compiled already
+     * @param string $potentialVariable candidate for a variable to seek in data
+     * @param array $data ['variable' => value]
+     * @return string string with a compiled variable
+     * @throws Exception if data doesn't include the variable
+     */
+    protected function recordVariable(string $compiledString, string $potentialVariable, array $data) : string
+    {
+        $potentialVariable = trim(ltrim($potentialVariable, self::DOLLAR_SIGN_TAG)); // makes {$ variable } and similar also allowed
+        if (array_key_exists($potentialVariable, $data)) {
+            if (filter_var($potentialVariable, FILTER_VALIDATE_URL)) {
+                return $this->handleUrl($compiledString, $potentialVariable, $data);
+            } else {
+                return $compiledString . $data[$potentialVariable];
+            }
+        }
+
+        throw new Exception('Variable with name ' . $potentialVariable . ' is not assigned to this email template');
```

Comment:
Again I think we'll need to handle this a little bit better for the user.

Reply:
Depending on how we're doing it, it might be best just to let the mistake go out with the email. I'd like to catch the exception and report back to the user, but that might be difficult since by the time the email is sent, in most cases the action is already finished.

So it might not be critical right now, but we might want to think about how we can make use of this code to support a validation step of some kind, so that a potential email could be validated even before the event takes place (for example, with reviewer assignment, the user should get an error in the UI and the assignment shouldn't happen if their email is not valid for some reason).